core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -28,41 +34,180 @@
  * @author iyzhang
  */
 public class KernelClient {
+    private static final int MIN_METRIC_POLL_PERIOD_MS = 1000; /* Minimum metrics poll period */
+    private static final int MAX_METRIC_POLL_PERIOD_MS = 512000; /* Maximum metrics poll period */
+    private static final int STEP_SIZE_IN_BYTES = 1024; /* Step size in bytes */
+    /* Random byte array used to send in heartbeats */
+    private static final byte[] randomBytes = new byte[STEP_SIZE_IN_BYTES];
+
+    static {
+        new Random().nextBytes(randomBytes);
+    }
+
+    /* Data holder class used in heartbeats */
+    public static class RandomData implements Serializable {
+        private transient int len = STEP_SIZE_IN_BYTES; /* length of data to be sent to server */
+
+        private void writeObject(java.io.ObjectOutputStream s) throws java.io.IOException {
+            s.writeInt(len);
+            for (int i = 0; i < len; i += STEP_SIZE_IN_BYTES) {
+                /* Write the same byte array again as this data is useless on receiver end */
+                s.write(randomBytes);
+            }
+        }
+
+        private void readObject(java.io.ObjectInputStream s)
+                throws java.io.IOException, ClassNotFoundException {
+            len = s.readInt();
+            for (int i = 0; i < len; i += STEP_SIZE_IN_BYTES) {
+                /* Received data is useless. Just overwrites to same byte array */
+                s.read(randomBytes);
+            }
+        }
+    }
+
     /** Stub for the OMS */
     private OMSServer oms;
-    /** List of hostnames matched to kernel server stubs */
-    private Hashtable<InetSocketAddress, KernelServer> servers;
+
+    /* Maximum number of continuous heartbeat miss allowed */
+    private static final int MAX_HEARTBEAT_MISS_TIMES = 3;
+
+    /* Class to hold remote kernel server info. This class is accessible only within this outer class(KernelClient) */
+    private final class KernelServerInfo {
+        private InetSocketAddress serverAddress; /* Host address of kernel server */
+        private KernelServer remoteRef; /* Remote reference to the kernel server */
+        /* Below reference counts are used to remove a server when it fails to send heartbeats for
+        MAX_HEARTBEAT_MISS_ALLOWED times continuously */
+        /* Global reference count is incremented whenever heartbeat is attempted */
+        private int refCount;
+        private int refCountOnHeartBeat; /* Reference count upon successful heartbeats */
+
+        /* Minimum samples to consider data rates as consistent. */
+        private static final int MIN_STABLE_DATA_RATE_TIMES = 10;
+        private int stableDataRateTimes; /* Number of continuous stable data transfer rates */
+        private RandomData data; /* Amount of random data sent to server in heartbeats */
```

Comment:
I don't think that all heartbeats should send data.  As discussed, data will often be very large and should therefore be sent as infrequently as possible (e.g. to ensure that this data does not consume more than say 1% of the link capacity).  Heartbeats on the other hand need to be cheap and sent frequently(e.g.  every few seconds?), so that automatic recovery from link failure can be fast.

So in summary I think that the rate of sending heartbeats, and the rate of sending large amounts of data to measure throughput should be decoupled.

Reply:
Actually, there is no heartbeat between kernel servers. It is just that method name is heartbeat(). But they are actually used to measure latency and data rate(i.e., not liveliness). And once we get the stable samples, we reduce the frequency of the measuring them gradually. 
We already have heartbeat mechanism between OMS and Kernel Server. So OMS can detect if a kernel server is not reachable and un-register that server if heartbeat is missed consecutively thrice. Note that, all DM group policy objects reside on OMS. So group policy can know if a server policy is not reachable.
So, If a remote kernel server is down due to some reason, any rpc calls to that particular server fails, then the client contacts respective DM group policy to get the available server, uses it.
So, IMO, we don't need precise liveliness detection from each server to all other remaining servers.

Reply:
If node A gets disconnected from node B, then microservices on node A that rely on/connect to microservices on node B will stop working.  Something needs to detect that and relocate the microservices from either node A or node B.

Reply:
Also, please don't call something "heartbeat" if it's not a heartbeat.  Heartbeats are regular and frequent (at least mine are :-)

Reply:
Renamed them as below:
![Selection_020](https://user-images.githubusercontent.com/35334869/63342605-84e39300-c369-11e9-8d08-2bdebd837082.png)